### ISSUE TITLE
DNS request timed out 4 times.

### ISSUE THREAD

**Issue by YanLemon** [kubernetes/dns#313]
I use k8s ambassador NodePort to expose serive. I open firewalld and then close. the ambassador pod log show DNS request timed out 4 times.
I kubectl exec ambassador pod:
/var/log # netstat -nal| grep udp | grep -v 127
udp 0 0 177.177.162.240:60021 10.96.0.10:53 ESTABLISHED

I want to ask: DNS request timed out 4 times, why not closeSocket?

[root@node1 ~]# kubectl get svc --all-namespaces
NAMESPACE NAME TYPE CLUSTER-IP EXTERNAL-IP PORT(S) AGE
base-service ambassador NodePort 10.96.140.102 443:10443/TCP,80:10080/TCP 1d
base-service cas-server LoadBalancer 10.96.3.228 3000:51487/TCP
kube-system kube-dns ClusterIP 10.96.0.10 53/UDP,53/TCP

[root@node1 opt]#systemctl stop firewalld

[root@node1 opt]# kubectl exec ambassador-ngvc9 -n base-service -ti -- /bin/sh

/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador #
/ambassador #
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy
/ambassador #
/ambassador #
/ambassador # netstat -nalp |grep udp | grep -v 127
udp 0 0 177.177.166.134:41963 10.96.0.10:53 ESTABLISHED 438/envoy

the pod log:
[2019-06-27 05:42:38.313] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds
[2019-06-27 05:42:38.313] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds
[2019-06-27 05:42:38.313] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds
[2019-06-27 05:42:38.313] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 3793 milliseconds
[2019-06-27 05:42:38.882] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats
[2019-06-27 05:42:39.883] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats
[2019-06-27 05:42:40.884] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats
[2019-06-27 05:42:41.659] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/upstream/upstream_impl.cc:866] starting async DNS resolution for 127.0.0.1
[2019-06-27 05:42:41.659] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/upstream/upstream_impl.cc:873] async DNS resolution complete for 127.0.0.1
[2019-06-27 05:42:41.885] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats
[2019-06-27 05:42:42.107] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:118] DNS request timed out 4 times
[2019-06-27 05:42:42.107] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/upstream/upstream_impl.cc:873] async DNS resolution complete for dnec-ui.dnec
[2019-06-27 05:42:42.107] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 2008 milliseconds
[2019-06-27 05:42:42.886] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats
[2019-06-27 05:42:43.149] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/connection_handler_impl.cc:217] [C42459] new connection
[2019-06-27 05:42:43.149] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:187] [C42459] new stream
[2019-06-27 05:42:43.149] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:798] [C42459][S2738831110447437719] request end stream
[2019-06-27 05:42:43.149] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:454] [C42459][S2738831110447437719] request headers complete (end_stream=true):
':authority', '127.0.0.1:8001'
':path', '/logging'
':method', 'POST'
'user-agent', 'python-requests/2.18.4'
'accept-encoding', 'gzip, deflate'
'accept', '/'
'connection', 'keep-alive'
'content-length', '0'

[2019-06-27 05:42:43.149] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 admin: source/server/http/admin.cc:755] [C42459][S2738831110447437719] request complete: path: /logging
[2019-06-27 05:42:43.149] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:984] [C42459][S2738831110447437719] encoding headers via codec (end_stream=false):
':status', '404'
'content-type', 'text/plain; charset=UTF-8'
'cache-control', 'no-cache, max-age=0'
'x-content-type-options', 'nosniff'
'date', 'Thu, 27 Jun 2019 05:42:43 GMT'
'server', 'envoy'

[2019-06-27 05:42:43.150] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 connection: source/common/network/connection_impl.cc:451] [C42459] remote close
[2019-06-27 05:42:43.150] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 connection: source/common/network/connection_impl.cc:133] [C42459] closing socket: 0
[2019-06-27 05:42:43.150] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/connection_handler_impl.cc:50] [C42459] adding to cleanup list
[2019-06-27 05:42:43.153] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/connection_handler_impl.cc:217] [C42460] new connection
[2019-06-27 05:42:43.153] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:187] [C42460] new stream
[2019-06-27 05:42:43.153] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:798] [C42460][S11881029595277275926] request end stream
[2019-06-27 05:42:43.153] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:454] [C42460][S11881029595277275926] request headers complete (end_stream=true):
':authority', '127.0.0.1:8001'
':path', '/stats'
':method', 'GET'
'user-agent', 'python-requests/2.18.4'
'accept-encoding', 'gzip, deflate'
'accept', '/'
'connection', 'keep-alive'

[2019-06-27 05:42:43.153] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 admin: source/server/http/admin.cc:755] [C42460][S11881029595277275926] request complete: path: /stats
[2019-06-27 05:42:43.158] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 http: source/common/http/conn_manager_impl.cc:984] [C42460][S11881029595277275926] encoding headers via codec (end_stream=false):
':status', '200'
'content-type', 'text/plain; charset=UTF-8'
'cache-control', 'no-cache, max-age=0'
'x-content-type-options', 'nosniff'
'date', 'Thu, 27 Jun 2019 05:42:43 GMT'
'server', 'envoy'

[2019-06-27 05:42:43.159] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 connection: source/common/network/connection_impl.cc:451] [C42460] remote close
[2019-06-27 05:42:43.159] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 connection: source/common/network/connection_impl.cc:133] [C42460] closing socket: 0
[2019-06-27 05:42:43.159] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/connection_handler_impl.cc:50] [C42460] adding to cleanup list
[2019-06-27 05:42:43.887] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats
[2019-06-27 05:42:44.114] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 1 milliseconds
[2019-06-27 05:42:44.117] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 upstream: source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 4288 milliseconds
[2019-06-27 05:42:44.887] ambassador-ngvc9 ambassador/debug/AMBSSADOR_LOG: 438 main: source/server/server.cc:119] flushing stats

**Comment by YanLemon** [comment]
@MrHohn @prameshj  Could you give me some advice? thank you.

**Comment by MrHohn** [comment]
@YanLemon Could you be more specific on what the issue is? I found it hard to parse out the actual info.

What is the exact DNS request failure (e.g. looking up what hostname)? What does "open firewalld and then close" mean? Is this issue generic or specific to the ambassador pod (maybe file an issue against https://github.com/datawire/ambassador instead)?

**Comment by YanLemon** [comment]
@MrHohn [root@node1 ~]# systemctl stop firewalld.service
[root@node1 ~]# systemctl start firewalld.service


[root@node1 ~]# tcpdump udp port 53 | grep 57129
16:20:37.743377 IP 177.177.135.3.domain > 177.177.166.134.57129: 21511 NXDomain 0/1/0 (154)
16:20:37.743458 IP 177.177.166.134.57129 > 177.177.135.3.domain: 55103+ A? dnec-ui.dnec.svc.cluster.local. (48)
16:20:37.743760 IP 177.177.135.3.domain > 177.177.166.134.57129: 55103* 1/0/0 A 10.96.28.29 (94)
16:20:42.377844 IP 177.177.166.134.57129 > 177.177.135.3.domain: 40513+ A? cas-server.base-service.svc.cluster.local. (59)
16:20:42.378244 IP 177.177.135.3.domain > 177.177.166.134.57129: 40513* 1/0/0 A 10.96.3.228 (116)
16:20:42.744768 IP 177.177.166.134.57129 > 177.177.135.3.domain: 25186+ A? dnec-ui.dnec.base-service.svc.cluster.local. (61)
16:20:42.745102 IP 177.177.135.3.domain > 177.177.166.134.57129: 25186 NXDomain 0/1/0 (154)
16:20:42.745171 IP 177.177.166.134.57129 > 177.177.135.3.domain: 53302+ A? dnec-ui.dnec.svc.cluster.local. (48)
16:20:42.745500 IP 177.177.135.3.domain > 177.177.166.134.57129: 53302* 1/0/0 A 10.96.28.29 (94)
16:20:45.972268 IP 177.177.166.134.57129 > 177.177.135.3.domain: 55922+ A? log-service.base-service.svc.cluster.local. (60)
16:20:45.972764 IP 177.177.135.3.domain > 177.177.166.134.57129: 55922* 1/0/0 A 10.96.94.81 (118)
16:20:46.751043 IP 177.177.166.134.57129 > 177.177.135.3.domain: 44478+ A? iam-mapping.base-service.svc.cluster.local. (60)
16:20:46.751288 IP 177.177.135.3.domain > 177.177.166.134.57129: 44478 1/0/0 A 10.96.243.131 (118)
16:20:47.380482 IP 177.177.166.134.57129 > 177.177.135.3.domain: 30885+ A? cas-server.base-service.svc.cluster.local. (59)
16:20:47.380981 IP 177.177.135.3.domain > 177.177.166.134.57129: 30885* 1/0/0 A 10.96.3.228 (116)
16:20:47.746540 IP 177.177.166.134.57129 > 177.177.135.3.domain: 44754+ A? dnec-ui.dnec.base-service.svc.cluster.local. (61)
16:20:47.746905 IP 177.177.135.3.domain > 177.177.166.134.57129: 44754 NXDomain 0/1/0 (154)
16:20:47.746960 IP 177.177.166.134.57129 > 177.177.135.3.domain: 22101+ A? dnec-ui.dnec.svc.cluster.local. (48)
16:20:47.747260 IP 177.177.135.3.domain > 177.177.166.134.57129: 22101* 1/0/0 A 10.96.28.29 (94)
16:20:50.564552 IP 177.177.166.134.57129 > 177.177.135.3.domain: 34747+ A? dnec-be.dnec.base-service.svc.cluster.local. (61)
16:20:50.564898 IP 177.177.135.3.domain > 177.177.166.134.57129: 34747 NXDomain 0/1/0 (154)
16:20:50.564963 IP 177.177.166.134.57129 > 177.177.135.3.domain: 17562+ A? dnec-be.dnec.svc.cluster.local. (48)
16:20:50.565110 IP 177.177.135.3.domain > 177.177.166.134.57129: 17562 1/0/0 A 10.96.9.110 (94)
16:20:50.874064 IP 177.177.166.134.57129 > 177.177.135.3.domain: 11220+ A? licc.base-service.svc.cluster.local. (53)
16:20:50.874520 IP 177.177.135.3.domain > 177.177.166.134.57129: 11220* 1/0/0 A 10.96.248.20 (104)
16:20:50.974310 IP 177.177.166.134.57129 > 177.177.135.3.domain: 56967+ A? log-service.base-service.svc.cluster.local. (60)
16:20:50.974843 IP 177.177.135.3.domain > 177.177.166.134.57129: 56967* 1/0/0 A 10.96.94.81 (118)
16:20:51.751794 IP 177.177.166.134.57129 > 177.177.135.3.domain: 14126+ A? iam-mapping.base-service.svc.cluster.local. (60)
16:20:51.752110 IP 177.177.135.3.domain > 177.177.166.134.57129: 14126 1/0/0 A 10.96.243.131 (118)
16:20:52.381871 IP 177.177.166.134.57129 > 177.177.135.3.domain: 34881+ A? cas-server.base-service.svc.cluster.local. (59)
16:20:52.382411 IP 177.177.135.3.domain > 177.177.166.134.57129: 34881* 1/0/0 A 10.96.3.228 (116)
16:20:52.747616 IP 177.177.166.134.57129 > 177.177.135.3.domain: 8107+ A? dnec-ui.dnec.base-service.svc.cluster.local. (61)
16:20:52.748297 IP 177.177.135.3.domain > 177.177.166.134.57129: 8107 NXDomain 0/1/0 (154)
16:20:52.748431 IP 177.177.166.134.57129 > 177.177.135.3.domain: 37011+ A? dnec-ui.dnec.svc.cluster.local. (48)
16:20:52.748751 IP 177.177.135.3.domain > 177.177.166.134.57129: 37011* 1/0/0 A 10.96.28.29 (94)
16:20:53.958038 IP 177.177.166.134.57129 > 177.177.135.3.domain: 61525+ A? dnec-deploy.dnec.base-service.svc.cluster.local. (65)
16:20:53.958354 IP 177.177.135.3.domain > 177.177.166.134.57129: 61525 NXDomain 0/1/0 (158)
16:20:53.958474 IP 177.177.166.134.57129 > 177.177.135.3.domain: 52290+ A? dnec-deploy.dnec.svc.cluster.local. (52)
16:20:53.958785 IP 177.177.135.3.domain > 177.177.166.134.57129: 52290* 1/0/0 A 10.96.108.86 (102)
16:20:54.936336 IP 177.177.166.134.57129 > 177.177.135.3.domain: 23618+ A? iam-service.base-service.svc.cluster.local. (60)
16:20:54.936788 IP 177.177.135.3.domain > 177.177.166.134.57129: 23618 1/0/0 A 10.96.118.55 (118)

**Comment by YanLemon** [comment]
1、start or stop firewalld，the dns lookup will resolve fail for a while ,and then back to normal.
2、start firewalld,the envoy will close the socket (for DNS request)
3、There are some phenomena when stop the firewalld:
1) stop the fiewalld, the same source port (same as the port before stop firewalld ) can receive DNS response
2) stop the fiewalld,the same port has no response from dns
3) stop the fiewalld,the envoy change the source port and has no response from dns
4、when envoy has no dns response, the socket never close (the source port never changed)

[root@node1 ~]# systemctl stop firewalld.service
[root@node1 ~]# systemctl start firewalld.service

**Comment by johnbelamaric** [comment]
@YanLemon sounds like you need to file this with envoy or ambassador. The DNS server has no relation to this client-side behavior.
